actionview/lib/action_view/helpers/form_tag_helper.rb

```diff
@@ -420,10 +420,18 @@ def text_area_tag(name, content = nil, options = {})
         content_tag :textarea, content.to_s.html_safe, { "name" => name, "id" => sanitize_to_id(name) }.update(options)
       end
 
+      ##
+      # :call-seq:
+      #   check_box_tag(name, options = {})
+      #   check_box_tag(name, value = 1, options = {})
+      #   check_box_tag(name, value = 1, checked = false, options = {})
```

Comment:
This is a bit confusing to me.

In https://github.com/rails/rails/pull/45527#discussion_r917276410, I was trying to allude to [method overloading](https://en.wikipedia.org/wiki/Function_overloading), such as in Java.  Arg default values are a kind of alternative to method overloading, though I think you can mix and match up to a point.  So the overloads I wrote in https://github.com/rails/rails/pull/45527#discussion_r917276410 would be expanded as:

* `check_box_tag(name, options = {})` =>
  * `check_box_tag(name, options)`
  * `check_box_tag(name)`
* `check_box_tag(name, value, options = {})` =>
  * `check_box_tag(name, value, options)`
  * `check_box_tag(name, value)`
* `check_box_tag(name, value, checked, options = {})` =>
  * `check_box_tag(name, value, checked, options)`
  * `check_box_tag(name, value, checked)`

Note that each expansion is unique, so if a compiler were trying to statically resolve the overload, it could do so.

On the other hand, the overloads written here would be expanded as:

* `check_box_tag(name, options = {})` =>
  * `check_box_tag(name, options)`
  * `check_box_tag(name)`
* `check_box_tag(name, value = 1, options = {})` =>
  * `check_box_tag(name, value, options)`
  * `check_box_tag(name, value)`
  * `check_box_tag(name)`
* `check_box_tag(name, value = 1, checked = false, options = {})` =>
  * `check_box_tag(name, value, checked, options)`
  * `check_box_tag(name, value, checked)`
  * `check_box_tag(name, value)`
  * `check_box_tag(name)`

Notice the overlaps.  Obviously, we don't have a compiler trying to statically resolve these, but we do some form of resolution in our heads.  At least, that's how I think about these things.

Reply:
Thanks. I understand a bit better what you were trying to say now.